tree-sitter-markdown/src/scanner.c

```diff
@@ -0,0 +1,1446 @@
+    ATX_H1_MARKER,
+    ATX_H2_MARKER,
+    ATX_H3_MARKER,
+    ATX_H4_MARKER,
+    ATX_H5_MARKER,
+    ATX_H6_MARKER,
+    SETEXT_H1_UNDERLINE,
+    SETEXT_H2_UNDERLINE,
+    THEMATIC_BREAK,
+    LIST_MARKER_MINUS,
+    LIST_MARKER_PLUS,
+    LIST_MARKER_STAR,
+    LIST_MARKER_PARENTHESIS,
+    LIST_MARKER_DOT,
+    LIST_MARKER_MINUS_DONT_INTERRUPT,
+    LIST_MARKER_PLUS_DONT_INTERRUPT,
+    LIST_MARKER_STAR_DONT_INTERRUPT,
+    LIST_MARKER_PARENTHESIS_DONT_INTERRUPT,
+    LIST_MARKER_DOT_DONT_INTERRUPT,
+    FENCED_CODE_BLOCK_START_BACKTICK,
+    FENCED_CODE_BLOCK_START_TILDE,
+    BLANK_LINE_START,
+    FENCED_CODE_BLOCK_END_BACKTICK,
+    FENCED_CODE_BLOCK_END_TILDE,
+    HTML_BLOCK_1_START,
+    HTML_BLOCK_1_END,
+    HTML_BLOCK_2_START,
+    HTML_BLOCK_3_START,
+    HTML_BLOCK_4_START,
+    HTML_BLOCK_5_START,
+    HTML_BLOCK_6_START,
+    HTML_BLOCK_7_START,
+    CLOSE_BLOCK,
+    NO_INDENTED_CHUNK,
+    ERROR,
+    TRIGGER_ERROR,
+    TOKEN_EOF,
+    MINUS_METADATA,
+    PLUS_METADATA,
+    PIPE_TABLE_START,
+    PIPE_TABLE_LINE_ENDING,
+} TokenType;
+
+// Description of a block on the block stack.
+//
+// LIST_ITEM is a list item with minimal indentation (content begins at indent level 2) while
+// LIST_ITEM_MAX_INDENTATION represents a list item with maximal indentation without being
+// considered a indented code block.
+//
+// ANONYMOUS represents any block that whose close is not handled by the external scanner.
+typedef enum {
+    BLOCK_QUOTE,
+    INDENTED_CODE_BLOCK,
+    LIST_ITEM,
+    LIST_ITEM_1_INDENTATION,
+    LIST_ITEM_2_INDENTATION,
+    LIST_ITEM_3_INDENTATION,
+    LIST_ITEM_4_INDENTATION,
+    LIST_ITEM_5_INDENTATION,
+    LIST_ITEM_6_INDENTATION,
+    LIST_ITEM_7_INDENTATION,
+    LIST_ITEM_8_INDENTATION,
+    LIST_ITEM_9_INDENTATION,
+    LIST_ITEM_10_INDENTATION,
+    LIST_ITEM_11_INDENTATION,
+    LIST_ITEM_12_INDENTATION,
+    LIST_ITEM_13_INDENTATION,
+    LIST_ITEM_14_INDENTATION,
+    LIST_ITEM_MAX_INDENTATION,
+    FENCED_CODE_BLOCK,
+    ANONYMOUS,
+} Block;
+
+// Determines if a character is punctuation as defined by the markdown spec.
+static bool is_punctuation(char c) {
+    return
+        (c >= '!' && c <= '/') ||
+        (c >= ':' && c <= '@') ||
+        (c >= '[' && c <= '`') ||
+        (c >= '{' && c <= '~');
+}
+
+// Returns true if the block represents a list item
+static bool is_list_item(Block block) {
+    return block >= LIST_ITEM && block <= LIST_ITEM_MAX_INDENTATION;
+}
+
+// Returns the indentation level which lines of a list item should have at minimum. Should only be
+// called with blocks for which `is_list_item` returns true.
+static uint8_t list_item_indentation(Block block) {
+    return block - LIST_ITEM + 2;
+}
+
+const size_t NUM_HTML_TAG_NAMES_RULE_1 = 3;
+const char* HTML_TAG_NAMES_RULE_1[NUM_HTML_TAG_NAMES_RULE_1] = { "pre", "script", "style" };
+const size_t NUM_HTML_TAG_NAMES_RULE_7 = 62;
+const char* HTML_TAG_NAMES_RULE_7[NUM_HTML_TAG_NAMES_RULE_7] = {
```

Comment:
I don't think defining array sizes with `const size_t` is compliant with the C standard. At the same time, I think treesitter expects `NUM_HTML_TAG_NAMES_RULE_1` and `NUM_HTML_TAG_NAMES_RULE_7` symbols to be actually there? So either an additional macro to initialize both, or inline value for the array size